src/Server/TCPHandler.h

```diff
@@ -231,24 +241,15 @@ class TCPHandler : public Poco::Net::TCPServerConnection
     std::optional<UInt64> nonce;
     String cluster;
 
-    /// `out_mutex` protects `out` (WriteBuffer).
-    /// So it is used for method sendData(), sendProgress(), sendLogs(), etc.
-    std::mutex out_mutex;
-    /// `task_callback_mutex` protects tasks callbacks.
-    /// Inside these callbacks we might also change cancellation status,
-    /// so it also protects cancellation status checks.
-    std::mutex task_callback_mutex;
-
-    /// At the moment, only one ongoing query in the connection is supported at a time.
-    QueryState state;
+    /// `callback_mutex` protects using `out` (WriteBuffer) and `in` (ReadBuffer) inside callback.
+    /// So it is used for method sendData(), sendProgress(), sendLogs() along with all callbacks which interact with that buffers
+    std::mutex callback_mutex;
```

Comment:
I'm not sure I totally understand. Don't we still need a mutex to protect `out` outside the callback?

Reply:
I decided to not hold the mutexs when I'm sure that there are no concurrent threads are run.
So I need the coordination inside callbacks and inside TCPHandler when TCPHandler acts concurrently with them.